### DISCUSSION TITLE
[@types/cacheable-request] Type Error is there (Cannot use namespace 'ResponseLike' as a type)

### DISCUSSION THREAD

**Discussion by baseballyama** [DefinitelyTyped/DefinitelyTyped#62099] [Issues with a @types package] [answered]
If I use `commonjs` at `compilerOptions.module` in `tsconfig.json`, I got type error when I run `npx tsc --noEmit`.

![image](https://user-images.githubusercontent.com/19153718/188255853-9b6da732-c6df-45d1-8ea4-ac2acf925105.png)

## REPL and details

https://github.com/baseballyama/cacheable-request-bug

[a comment by an automated account is not shown]

**Comment by BendingBender** [comment, marked as answer]
You don't need `@types/cacheable-request` with `cacheable-request@9`. The package provides it's own types. Remove the declarations and install the `@types/node` package to use them:

```sh
npm remove @types/cacheable-request @types/responselike && npm i -D @types/node
```

See also: https://github.com/DefinitelyTyped/DefinitelyTyped/pull/62102

**Reply by Rehan-05** [reply]
But I have one question, I did not install @/types/casheable-request or @types/resonsiblelike. It automatically install  when we install the node module. 
But I run this command npm remove @types/cacheable-request @types/responselike && npm i -D @types/node but same error occuring.

**Comment by Rehan-05** [comment]
Same error. Here is Package.json file of my Node js Project. Please check the type version in it and tell me about any solution. Thanks

{
  "name": "functions",
  "scripts": {
    "lint": "eslint --ext .js,.ts .",
    "build": "tsc",
    "build:watch": "tsc --watch",
    "serve": "npm run build && firebase emulators:start --only functions",
    "shell": "npm run build && firebase functions:shell",
    "dev": "npm run build && firebase serve",
    "start": "npm run shell",
    "deploy": "firebase deploy --only functions",
    "logs": "firebase functions:log"
  },
  "engines": {
    "node": "16"
  },
  "main": "lib/index.js",
  "dependencies": {
    "axios": "^0.27.2",
    "cookie-parser": "^1.4.6",
    "cron": "^2.1.0",
    "dotenv": "^16.0.1",
    "express": "^4.18.1",
    "firebase-admin": "^10.0.2",
    "firebase-functions": "^3.18.0",
    "jsonwebtoken": "^8.5.1",
    "node-cron": "^3.0.2",
    "oauth": "^0.10.0",
    "qs": "^6.11.0",
    "util": "^0.12.4",
    "uuid": "^8.3.2"
  },
  "devDependencies": {
    **"@types/cacheable-request": "^6.0.2",
    "@types/node": "^18.7.15",
    "@types/responselike": "^1.0.0",**
    "@typescript-eslint/eslint-plugin": "^5.12.0",
    "@typescript-eslint/parser": "^5.12.0",
    "eslint": "^8.9.0",
    "eslint-config-google": "^0.14.0",
    "eslint-plugin-import": "^2.25.4",
    "firebase-functions-test": "^0.2.0",
    "typescript": "^4.5.4"
  },
  "private": true
}

**Reply by BendingBender** [reply]
I really don't get why you need to install these types at all, none of the packages in your `package.json` rely on these types.

Nevertheless, just for the sake of completeness, if you use `npm` (>8) you can deeply [override](https://docs.npmjs.com/cli/v8/configuring-npm/package-json#overrides) the version of this dependency via the following entry in your `package.json`:
```json
  "overrides": {
    "@types/cacheable-request": "^6.0.2",
    "@types/responselike": "^1.0.0"
  }
```

If you use `yarn` you can add [resolutions](https://classic.yarnpkg.com/lang/en/docs/selective-version-resolutions) to your `package.json` to achieve the same result:
```json
  "resolutions": {
    "@types/cacheable-request": "^6.0.2",
    "@types/responselike": "^1.0.0"
  }
```

**Comment by nbouvrette** [comment]
None of these workarounds worked for me. Until the [root of the issue](https://github.com/sindresorhus/got/issues/2138) is resolved, this one did the trick: (requires npm [v8.3.0](https://github.com/npm/cli/releases/tag/v8.3.0) (2021-12-09)):

Update your `package.json` and add:

```json
  "overrides": {
    "@types/cacheable-request": "8.3.1"
  },
```

Edit `/node_modules/cacheable-request/dist/types.d.ts` and remove the 4 first lines `/// <reference types="node" resolution-mode="require"/>` (they are triggering a `'resolution-mode' assertions are only supported when 'moduleResolution' is 'node16' or 'nodenext'.` error and I could not find anything on Google... removing them worked for me)

**Comment by gklasen** [comment]
Nothing worked for me, but @nbouvrettes solution with slight modification:

Instead of `"@types/cacheable-request": "8.3.1"` in the overrides-section, i did `"got": "12.4.1"`. 
The error with moduleResolution was "skipped" by setting `"skipLibCheck": true` in tsconfig.json instead of deleting the lines.

**Reply by lucasassisrosa** [reply]
Same for me. When adding `"@types/cacheable-request": "8.3.1"` to `package.json` overrides, I got the following error

```
11.33 error TS2688: Cannot find type definition file for 'cacheable-request'.
11.33   The file is in the program because:
11.33     Entry point for implicit type library 'cacheable-request'
```

now with 

```json
"overrides": {
  "got": "12.4.1"
},
```

it worked
